pkg/controller/accountclaim/organizational_units.go

```diff
@@ -121,8 +123,15 @@ func MoveAccount(reqLogger logr.Logger, client awsclient.Client, account *awsv1a
 				if check {
 					return awsv1alpha1.ErrAccAlreadyInOU
 				}
+			case "ConcurrentModificationException":
+				// if we encounter a race condition we simply try again after a random sleep interval
+				r := rand.Intn(5)
+				ConcurrentModificationExceptionMsg := fmt.Sprintf("CreateOrganizationalUnit:ConcurrentModificationException: Race condition while attempting to move Account: %s to OU: %s, trying again in %d seconds", account.Spec.AwsAccountID, OUID, r)
+				reqLogger.Info(ConcurrentModificationExceptionMsg)
+				time.Sleep(time.Duration(r) * time.Second)
+				MoveAccount(reqLogger, client, account, OUID, parentID)
```

Comment:
Slight concern (very unlikely) is if ConcurrentModificationException is returned on consecutive runs then we get recursion. Eventually it should all return but there is a possibility of it happening.

Reply:
@fahlmant I mean we will get recursion any time we run into the error but I wasn't be concerned since the chances of it taking say 30 seconds are 1/15625

Reply:
I don't think we should be retrying here. The "best practice" would be to have the main reconcile requeue (possibly with a delay) so the retry happens on the next run through.

Obv that would rely on the code path through this point being nicely idempotent, which I didn't dig into. But in theory, that next reconcile wouldn't reach this point because whatever operation this bounced off of would have already done the right thing. If we could count on that being the case, we could even just ignore this error and carry on.

Reply:
> But in theory, that next reconcile wouldn't reach this point because whatever operation this bounced off of would have already done the right thing.

@2uasimojo Yeah, that was the one of the things I was wondering about while writing this. I need to check that it is actually the case. @fahlmant I am wondering though, are there no other places where these errors could be occurring without our knowledge?

Reply:
Which errors? We log every AWS error we hit. If we know of one we should be handling different, then we should add it to this case statement. Otherwise, they go to default so they are still logged.

Reply:
And in the default case, the error percolates [up|https://github.com/openshift/aws-account-operator/pull/411/files#diff-90b216857326e9f99796eb1d9cefbc91R65] and will [requeue|https://github.com/openshift/aws-account-operator/blob/d849220c9a47ef7f2c32732838a87433f7957b6b/pkg/controller/accountclaim/accountclaim_controller.go#L336]. Which, assuming we're properly idempotent, is what we should be doing anyway.

Was there an actual problem caused by this error as observed? If it was just a case of, "hey, we saw this message and it made us worry," then we may not want to bother doing anything. But if it did cause a problem, IMO we should be focusing on idempotence, not special-casing.

Reply:
@2uasimojo @fahlmant I did some research and it is indeed the case that accounts get moved even when the error occurs so I am updating the PR to reflect that. I am, however, still catching this error in order to stop these messages from cluttering the logs:

`
{"level":"error","ts":1594987296.3907142,"logger":"kubebuilder.controller","msg":"Reconciler error","controller":"accountclaim-controller","request":"uhc-staging-1efka7k6ue5idru7t9u7ftd0i0c4tua3/ci-cluster-4-4","error":"ConcurrentModificationException: AWS Organizations can't complete your request because it conflicts with another attempt to modify the same entity. Try again later.\n\tstatus code: 400, request id: 1359eaf3-3acd-406f-80bf-8b7555d4ccf7","stacktrace":"github.com/openshift/aws-account-operator/vendor/github.com/go-logr/zapr.(*zapLogger).Error\n\tsrc/github.com/openshift/aws-account-operator/vendor/github.com/go-logr/zapr/zapr.go:128\ngithub.com/openshift/aws-account-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\tsrc/github.com/openshift/aws-account-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:217\ngithub.com/openshift/aws-account-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1\n\tsrc/github.com/openshift/aws-account-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158\ngithub.com/openshift/aws-account-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\tsrc/github.com/openshift/aws-account-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133\ngithub.com/openshift/aws-account-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil\n\tsrc/github.com/openshift/aws-account-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134\ngithub.com/openshift/aws-account-operator/vendor/k8s.io/apimachinery/pkg/util/wait.Until\n\tsrc/github.com/openshift/aws-account-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88"}
`